test/exercises/diffusion/heat_2d_dist_buffers.chpl

```diff
@@ -0,0 +1,149 @@
+/*
+  A distributed 2D finite-difference heat/diffusion equation solver
+
+  Computation is executed over a 2D distributed array.
+  The array distribution is managed by the `Block` distribution.
+  Tasks are spawned manually with a `coforall` loop and synchronization
+  is done manually using a `barrier`. Halo regions are shared across
+  locales manually via halo/buffer arrays.
+*/
+
+import BlockDist.Block,
+       Collectives.barrier,
+       Time.stopwatch;
+
+// create a stopwatch to time kernel execution
+var t = new stopwatch();
+config const writeTime = false;
+
+// compile with `-sRunCommDiag=true` to see comm diagnostics
+use CommDiagnostics;
+config param RunCommDiag = false;
+
+// declare configurable constants with default values
+config const nx = 256,      // number of grid points in x
+             ny = 256,      // number of grid points in y
+             nt = 50,       // number of time steps
+             alpha = 0.25;  // diffusion constant
+
+// define distributed domains and block-distributed array
+const indices = {0..<nx, 0..<ny},
+      indicesInner = indices.expand(-1),
+      INDICES = Block.createDomain(indices);
+var u: [INDICES] real;
+
+// apply initial conditions
+u = 1.0;
+u[nx/4..nx/2, ny/4..ny/2] = 2.0;
+
+// a type for creating a "skyline" array of halo buffers
+record haloArray {
+  var d: domain(1);
+  var v: [d] real;
+
+  proc init() do this.d = {0..0};
+  proc init(r: range(int)) do
+    this.d = {r};
+}
+
+// set up array of halo buffers over same distribution as 'u.targetLocales'
+var LOCALE_DOM = Block.createDomain(u.targetLocales().domain);
+var haloArrays: [LOCALE_DOM] [0..<4] haloArray;
+
+// buffer edge indices: North, East, South, West
+param N = 0, S = 1, E = 2, W = 3;
+
+// number of tasks per dimension based on Block distribution's decomposition
+const tidXMax = u.targetLocales().dim(0).high,
+      tidYMax = u.targetLocales().dim(1).high;
+
+// barrier for one task per locale
+var b = new barrier(numLocales);
+
+proc main() {
+  if RunCommDiag then startCommDiagnostics();
+
+  // spawn one task for each locale
+  t.start();
+  coforall (loc, (tidX, tidY)) in zip(u.targetLocales(), LOCALE_DOM) {
+    // run initialization and computation on the task for this locale
+    on loc {
```

Comment:
I think this could just be written `forall (tidX, tidY) in LOCALE_DOM` given that you've distributed `LOCALE_DOM`—i.e., remove the complexity of the zippering and the on-clause.

This also leans more into the theme of leaning on distributed domains/arrays for this implementation rather than writing an MPI-style SPMD computation with local arrays/tasks that happen to live in a global address space.

Reply:
Don't I need to assert that I create exactly one task per locale/iteration for the barrier synchronization to work though?

Reply:
I believe it'll work fine without the assertion.  You could add the assertion to be defensive, though there are arguably other assumptions in the program that might want to similarly be protected by assertions.  I'm not sure how the assertion would ever not pass, though I could be missing something.  Specifically, the Block distribution should never fail to create a task per locale, and should never use more than one task per locale for a domain that only has a single index per locale (fingers crossed as I type that).

Reply:
> Specifically, the Block distribution should never fail to create a task per locale

Okay, that's what I was missing. I'll switch it to a `forall` loop

Reply:
You should definitely check my work, but I'll be shocked (and sad!) if I'm wrong.

Reply:
This worked, thanks!